pflag/pflag.go

```diff
@@ -0,0 +1,428 @@
+	mapStringStringSlice = reflect.MapOf(reflect.TypeOf(""), stringSlice)
+	mapStringString      = reflect.MapOf(reflect.TypeOf(""), reflect.TypeOf(""))
+	stringSet            = reflect.MapOf(reflect.TypeOf(""), reflect.TypeOf(struct{}{}))
+
+	_ dials.Source = (*Set)(nil)
+)
+
+const (
+	dialsPFlagTag      = "dialspflag"
+	dialsPFlagShortTag = "dialspflagshort"
+	// HelpTextTag is the name of the struct tags for flag descriptions
+	HelpTextTag = "dialsdesc"
+	// DefaultFlagHelpText is the default help-text for fields with an
+	// unset dialsdesc tag.
+	DefaultFlagHelpText = "unset description (`" + HelpTextTag + "` struct tag)"
+)
+
+// NameConfig defines the parameters for separating components of a flag-name
+type NameConfig struct {
+	// FieldNameEncodeCasing is for the field names used by the flatten mangler
+	FieldNameEncodeCasing caseconversion.EncodeCasingFunc
+	// TagEncodeCasing is for the tag names used by the flatten mangler
+	TagEncodeCasing caseconversion.EncodeCasingFunc
+}
+
+// TODO(@sachi): update FieldNameEncodeCasing to EncodeGoCamelCase once it exists
+
+// DefaultFlagNameConfig defines a reasonably-defaulted NameConfig for field names
+// and tags
+func DefaultFlagNameConfig() *NameConfig {
+	return &NameConfig{
+		FieldNameEncodeCasing: caseconversion.EncodeUpperCamelCase,
+		TagEncodeCasing:       caseconversion.EncodeKebabCase,
+	}
+}
+
+func ptrified(template interface{}) (reflect.Value, reflect.Type, error) {
+	val := reflect.ValueOf(template)
+	if val.Kind() != reflect.Ptr {
+		return reflect.Value{}, nil, fmt.Errorf("non-pointer-type passed: %s", val.Type())
+	}
+
+	val = val.Elem()
+	if val.Kind() != reflect.Struct {
+		return reflect.Value{}, nil, fmt.Errorf("pointer-to-non-struct-type passed: %s", val.Type())
+	}
+	typ := val.Type()
+	out := ptrify.Pointerify(typ, val)
+	return val, out, nil
+}
+
+// NewCmdLineSet registers flags for the passed template value in the standard
+// library's main flag.CommandLine FlagSet so binaries using dials for flag
+// configuration can play nicely with libraries that register flags with the
+// standard library. (or libraries using dials can register flags and let the
+// actual process's Main() call Parse())
+func NewCmdLineSet(cfg *NameConfig, template interface{}) (*Set, error) {
+	pval, ptyp, ptrifyErr := ptrified(template)
+	if ptrifyErr != nil {
+		return nil, ptrifyErr
+	}
+
+	s := Set{
+		Flags:           pflag.CommandLine,
+		ParseFunc:       func() error { pflag.Parse(); return nil },
+		ptrType:         ptyp,
+		flagsRegistered: true,
+		NameCfg:         cfg,
+		flagFieldName:   map[string]string{},
+		flagValues:      map[string]reflect.Value{},
+	}
+
+	if err := s.registerFlags(pval, ptyp); err != nil {
+		return nil, err
+	}
+
+	return &s, nil
+}
+
+// NewSetWithArgs creates a new pflag FlagSet and registers flags in it
+func NewSetWithArgs(cfg *NameConfig, template interface{}, args []string) (*Set, error) {
```

Comment:
Hmm, thinking about the `cobra` use-case, we'll want to pass the `PersistentFlags()` `pflag.FlagSet` for the subcommand as an argument to a constructor.
Let's add another constructor like this one that replaces `args` with a `FlagSet` arg (they both can call the same thing later), although the cobra use-case should set `ParseFunc` to `nil`.

Reply:
Added it. The helper function makes the pflag package a lot cleaner